### PR TITLE
[Clang][AST] fix crash in mangle lambda expression

### LINKED ISSUE #78542: clang crash when mangle lambda expression in lambda expression
https://godbolt.org/z/6M53v9njb
```c++
auto ICE = [](auto a) { return [=]<decltype(a) b>() { return 1; }; };
```

```
./build/bin/clang a.cpp -std=c++20
```

```
Assertion failed: (parmDepth < FunctionTypeDepth.getDepth()), function mangleFunctionParam, file ItaniumMangle.cpp, line 5693.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.

1.      ID/a.cpp:1:53: current parser token '{'
2.      ID/a.cpp:1:12: lambda expression parsing
3.      ID/a.cpp:1:23: in compound statement ('{}')
4.      ID/a.cpp:1:32: lambda expression parsing
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  clang-18                 0x00000001034397f0 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 88
1  clang-18                 0x0000000103439dd4 PrintStackTraceSignalHandler(void*) + 28
2  clang-18                 0x0000000103437abc llvm::sys::RunSignalHandlers() + 152
3  clang-18                 0x000000010343aec0 SignalHandler(int) + 276
4  libsystem_platform.dylib 0x000000018b991a24 _sigtramp + 56
5  libsystem_pthread.dylib  0x000000018b961cc0 pthread_kill + 288
6  libsystem_c.dylib        0x000000018b86da40 abort + 180
7  libsystem_c.dylib        0x000000018b86cd30 err + 0
8  clang-18                 0x0000000109dc5c7c (anonymous namespace)::CXXNameMangler::mangleFunctionParam(clang::ParmVarDecl const*) + 140
9  clang-18                 0x0000000109dc4ec8 (anonymous namespace)::CXXNameMangler::mangleExpression(clang::Expr const*, unsigned int, bool)::$_2::operator()(clang::NamedDecl const*) const + 216
10 clang-18                 0x0000000109dc1b8c (anonymous namespace)::CXXNameMangler::mangleExpression(clang::Expr const*, unsigned int, bool) + 9664
11 clang-18                 0x0000000109dbc5a8 (anonymous namespace)::CXXNameMangler::mangleType(clang::DecltypeType const*) + 212
12 clang-18                 0x0000000109db08bc (anonymous namespace)::CXXNameMangler::mangleType(clang::QualType) + 1020
13 clang-18                 0x0000000109dc7784 (anonymous namespace)::CXXNameMangler::mangleTemplateParamDecl(clang::NamedDecl const*) + 552
14 clang-18                 0x0000000109dccaf4 (anonymous namespace)::CXXNameMangler::mangleLambdaSig(clang::CXXRecordDecl const*) + 144
15 clang-18                 0x0000000109dac560 (anonymous namespace)::ItaniumMangleContextImpl::mangleLambdaSig(clang::CXXRecordDecl const*, llvm::raw_ostream&) + 88
16 clang-18                 0x0000000109da73f0 (anonymous namespace)::ItaniumNumberingContext::getManglingNumber(clang::CXXMethodDecl const*) + 168
17 clang-18                 0x00000001088cd68c clang::Sema::handleLambdaNumbering(clang::CXXRecordDecl*, clang::CXXMethodDecl*, std::__1::optional<clang::CXXRecordDecl::LambdaNumbering>) + 568
18 clang-18                 0x00000001088d1e18 clang::Sema::ActOnStartOfLambdaDefinition(clang::LambdaIntroducer&, clang::Declarator&, clang::DeclSpec const&) + 1276
19 clang-18                 0x0000000107932138 clang::Parser::ParseLambdaExpressionAfterIntroducer(clang::LambdaIntroducer&) + 4204
20 clang-18                 0x000000010792ff94 clang::Parser::ParseLambdaExpression() + 164
21 clang-18                 0x000000010791d970 clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, bool&, clang::Parser::TypeCastState, bool, bool*) + 19408
22 clang-18                 0x0000000107915acc clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, clang::Parser::TypeCastState, bool, bool*) + 92
23 clang-18                 0x00000001079143c0 clang::Parser::ParseAssignmentExpression(clang::Parser::TypeCastState) + 292
24 clang-18                 0x0000000107914268 clang::Parser::ParseExpression(clang::Parser::TypeCastState) + 36
25 clang-18                 0x00000001079a7e84 clang::Parser::ParseReturnStatement() + 720
26 clang-18                 0x00000001079a2374 clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::ParsedAttributes&, clang::ParsedAttributes&) + 2076
27 clang-18                 0x00000001079a19e8 clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) + 232
28 clang-18                 0x00000001079a980c clang::Parser::ParseCompoundStatementBody(bool) + 1028
29 clang-18                 0x00000001079321d0 clang::Parser::ParseLambdaExpressionAfterIntroducer(clang::LambdaIntroducer&) + 4356
30 clang-18                 0x000000010792ff94 clang::Parser::ParseLambdaExpression() + 164
31 clang-18                 0x000000010791d970 clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, bool&, clang::Parser::TypeCastState, bool, bool*) + 19408
32 clang-18                 0x0000000107915acc clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, clang::Parser::TypeCastState, bool, bool*) + 92
33 clang-18                 0x00000001079143c0 clang::Parser::ParseAssignmentExpression(clang::Parser::TypeCastState) + 292
34 clang-18                 0x00000001078bdb68 clang::Parser::ParseInitializer() + 56
35 clang-18                 0x00000001078bd08c clang::Parser::ParseDeclarationAfterDeclaratorAndAttributes(clang::Declarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::ForRangeInit*) + 1992
36 clang-18                 0x00000001078bb0b8 clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::ParsedAttributes&, clang::SourceLocation*, clang::Parser::ForRangeInit*) + 1976
37 clang-18                 0x00000001079c94c0 clang::Parser::ParseDeclOrFunctionDefInternal(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec&, clang::AccessSpecifier) + 1612
38 clang-18                 0x00000001079c89f4 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier) + 244
39 clang-18                 0x00000001079c82c8 clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) + 3468
40 clang-18                 0x00000001079c62e8 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) + 2032
41 clang-18                 0x00000001079c5a54 clang::Parser::ParseFirstTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) + 64
42 clang-18                 0x00000001078a120c clang::ParseAST(clang::Sema&, bool, bool) + 476
43 clang-18                 0x0000000104b46fe0 clang::ASTFrontendAction::ExecuteAction() + 300
44 clang-18                 0x000000010442df68 clang::CodeGenAction::ExecuteAction() + 100
45 clang-18                 0x0000000104b467a8 clang::FrontendAction::Execute() + 124
46 clang-18                 0x0000000104a661f8 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) + 824
47 clang-18                 0x0000000104c63e8c clang::ExecuteCompilerInvocation(clang::CompilerInstance*) + 1056
48 clang-18                 0x00000001007b87d4 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) + 1268
49 clang-18                 0x00000001007a6ff8 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) + 400
50 clang-18                 0x00000001007a5d24 clang_main(int, char**, llvm::ToolContext const&) + 648
51 clang-18                 0x00000001007e1efc main + 108
52 dyld                     0x000000018b5e10e0 start + 2360
clang: error: unable to execute command: Abort trap: 6
clang: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 18.0.0git (https://github.com/llvm/llvm-project.git f6617091a982c0802e9b980f2ce7e11a1355c38b)
Target: arm64-apple-darwin23.2.0
Thread model: posix
clang: note: diagnostic msg: 
********************
```

### PR DESCRIPTION
**Pull request by jcsxky**
Try to fix [issue](https://github.com/llvm/llvm-project/issues/78542) by creating a new `FunctionTypeDepthState` and pop it after finish.

### CODE REVIEW

File: clang/test/AST/mangle-lambda-expression-no-crash.cpp [hunk 1]
```diff
@@ -0,0 +1,4 @@
+// RUN: %clang_cc1 -fsyntax-only -std=c++20 -verify %s
```
Comment (cor3ntin) [inline clang/test/AST/mangle-lambda-expression-no-crash.cpp:1]: `-fsyntax-only` is suspicious, i would prefer a CodeGenTest that checks the mangle name
Maybe moving the test to `test/CodeGenCXX/mangle-lambdas.cpp ` (or one of the other similar test in this folder) would be a good idea

**Comment by cor3ntin** [comment]
Thanks for this patch.
Can you add a release note?

Does this fix any of these issues ? (if so, can you add test for them?

#76481 
#70064
#63271 
#61939
#45441

File: clang/test/AST/mangle-lambda-expression-no-crash.cpp [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+// expected-no-diagnostics
+
+auto ICE = [](auto a) { return [=]<decltype(a) b>() { return 1; }; };
```
Comment (cor3ntin) [inline clang/test/AST/mangle-lambda-expression-no-crash.cpp:4]: Can you reference the issue number in the test?

File: clang/test/AST/mangle-lambda-expression-no-crash.cpp (hunk 1 above)
Comment (jcsxky) [reply clang/test/AST/mangle-lambda-expression-no-crash.cpp]: Sorry, I am afraid that I misunderstand root cause of this issue and still need more working on it.

**Review by HerrCai0907 — COMMENTED** [review]
Should it change the mangle part in libcxx also?

**Comment by jcsxky** [comment]
@cor3ntin I test this code
```cpp
inline auto XXXYYY = [](auto a) {
	return [=]<typename b>() {
		return 1;
	};
};

void foo(){
	XXXYYY(1);
}

```
after name mangling, clang outputs:
```cpp
call void @_ZNK6XXXYYYMUlT_E_clIiEEDaS0_(ptr noundef nonnull align 1 dereferenceable(1) @XXXYYY, i32 noundef 1)
```
and gcc output:
```cpp
call	_ZNK6XXXYYYMUlT_E_clIiEEDaS_
```
Is this expected?

**Comment by cor3ntin** [comment]
both compiler seems to mangkle it as `_ZNK6XXXYYYMUlT_E_clIiEEDaS0_` which looks correct
(`auto XXXYYY::{lambda(auto:1)#1}::operator()<int>(int) const`) @rjmccall 

Compiler explorer let you demangle identifiers (in the output menu, under the compilers dropdown https://compiler-explorer.com/z/Th11TjM8K )

**Pull request closed without merging**